docs/spec/identities.md

```diff
@@ -0,0 +1,438 @@
+remote refs. If these preconditions are not met, the clone is aborted, and
+already fetched data is pruned.
+
+When **fetching** (i.e. updating) a repository, a predetermined set of verified
+delegates is already known -- their views of the attestation history SHALL be
+fetched first, along with any advertised histories not yet present in the local
+repository (as those might be owned by future delegates). The fetching end MUST
+now find a valid `Doc` revision near the tips of the available delegate
+histories, and pick the most recent one. If two valid identities are found of
+which neither is in the ancestry path of the other, the attestation chain is
+said to have **forked**. In lieu of a consensus system, it is undecidable which
+side of the fork to commit to, and so replication of the attested repository
+SHALL refuse any further updates.
+
+> **TODO** Fork detection (ie. pairwise compare if in same ancestry path, either
+> select most recent or flag as forked)
+
+## Git Encoding
+
+In the `git` implementation, a `Doc` corresponds to a `blob` object, stored as
+the single entry of a `tree` object, such that its name (acc. to the `tree`) is
+equal to the `blob` hash of the _initial_ version of the `Doc`, serialised in
+canonical form. That is:
+
+    let name = git hash-object -t blob doc.canonical_form()
+
+An `Identity` corresponds to a `commit` object.
+
+We map the fields as follows:
+
+```rust
+/* Simplified git object model */
+struct Commit {
+    id: Oid,
+    tree: Tree,
+    message: String,
+}
+
+struct Tree {
+    id: Oid,
+    entries: Vec<TreeEntry>,
+}
+
+struct TreeEntry {
+    id: Oid,
+    name: String,
+    object: Blob, // faithfully
+}
+
+struct Blob {
+    id: Oid,
+    content: Vec<u8>,
+}
+
+/* Mapping (trivial type conversions elided) */
+let commit = /* .. */;
+let identity = Identity {
+    id: commit.id,
+    root: commit.tree.entries[0].name,
+    revision: tree.id,
+    doc: deserialize(commit.tree.entries[0].blob.content),
+    signatures: from_trailers(commit.message),
+};
+```
+
+Where:
+
+* `deserialize` is implemented by a standard JSON parser
+* `from_trailers` interprets the commit message as per [git-interpret-trailers],
+  and extracts the signatures from trailers with the token `x-rad-signature`.
+
+> **TODO** Specify exact format of signature trailer value (BASE64(public key ||
+> signature))
+
+The commit chain is stored in a branch at `refs/rad/id`.
+
+> **TODO** Insert repository layout spec as per
+> [RFC](../rfc/identity_resolution.md) here
+
+### Workflow
+
+Our construction of the `Identity` from a git commit allows for multiple `id`s
+to describe the same revision of the document (and thus be equally valid). This
+means that the respective delegates' histories may diverge in their _commit_
+histories, but still converge to an agreement on the validity of the attested
+document revision.
+
+While this allows for arbitrarily complex workflows (and history topologies), we
```

Comment:
Would it be possible to specify a working workflow, and possibly mandate it?
Extensibility is good but leaving too much freedom in this direction can mean asking for trouble in the long run.

Dealing with the logical (or physical) merge of the commit histories of different devices is not trivial and we should provide a detailed spec of how this should work.

Reply:
I can detail it more, but it is no more complex than walking the history in topological order and skipping over commits which do not satisfy the **Quorum** rule. Recall that we are verifying the histories of the different `PeerId`s one-by-one, and have established a linear ancestry of _document_ revisions.

Reply:
So, if I got it right, the idea is to have actual merges between the different devices' branches.

If we have two devices D1 and D2 with keys K1 and K2, a user could

- define the identity on D1, sign it on K1, now it is part of D1's branch history but without a quorum
- merge it from D1 into D2's branch, still no quorum
- sign it there with key K2, now there is a valid entity with a quorum
- merge it back into D1 so that also D1 can see the entity with the quorum of signatures

The histories of D1 and D2 would contain various merge commits.
When walking them backwards care should be taken to pick the right parent if one wants to stay on the same device branch and not switch into the other.

It should not be that hard because commits in D1 are signed with K1 (and K2 on D2) but this logic must be implemented.

Is this the workflow we are aiming to?

And BTW, how do we create a second device for a user U to add a new key?

User U has device D1 with key K1.
The second device is supposed to have K2 as key but K2 is still not part of the user identity U (it cannot be: the user identity U is on D1 but D1 cannot sign it with K2).

Device D2 is supposed to have U as "self identity" but user U does not have the device key K2, so we have the impasse of a device with a key, but its "self" identity does not have that key as trusted (yet), so it cannot communicate with others and be verified (can it?).

It is these kind of passages that we should specify.

Reply:
I have no idea why you think this is complicated, and I don't even understand your questions. The `tree` stays the same for every revision, it is only the signatures in a given commit which determine whether that revision is accepted -- if you know enough signatures over that tree, you can include all of them in a new commit to pass the rules. The identity (commit hash) of that commit doesn't matter, and there can exist more than one across the different maintainer branches. Iow, whether you merge or not is irrelevant, it is also irrelevant which parent comes first if you do.

I added a simple example as a recommendation, but as I wrote it can be made almost arbitrarily complicated.